src/crypto/equihash.cpp

```diff
@@ -0,0 +1,287 @@
+// Copyright (c) 2016 Jack Grigg
+// Copyright (c) 2016 The Zcash developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+// Implementation of the Equihash Proof-of-Work algorithm.
+//
+// Reference
+// =========
+// Alex Biryukov and Dmitry Khovratovich
+// Equihash: Asymmetric Proof-of-Work Based on the Generalized Birthday Problem
+// NDSS ’16, 21-24 February 2016, San Diego, CA, USA
+// https://www.internetsociety.org/sites/default/files/blogs-media/equihash-asymmetric-proof-of-work-based-generalized-birthday-problem.pdf
+
+#include "crypto/equihash.h"
+#include "util.h"
+
+#include <algorithm>
+#include <cmath>
+#include <iostream>
+#include <stdexcept>
+
+void validate_params(int n, int k)
+{
+    if (k>=n) {
+        std::cerr << "n must be larger than k\n";
+        throw invalid_params();
+    }
+    if (n % 8 != 0) {
+        std::cerr << "Parameters must satisfy n = 0 mod 8\n";
+        throw invalid_params();
+    }
+    if ((n/(k+1)) % 8 != 0) {
+        std::cerr << "Parameters must satisfy n/(k+1) = 0 mod 8\n";
+        throw invalid_params();
+    }
+}
+
+int Equihash::InitialiseState(eh_HashState& base_state)
+{
+    unsigned char personalization[crypto_generichash_blake2b_PERSONALBYTES] = {};
+    memcpy(personalization, "ZcashPOW", 8);
+    memcpy(personalization+8,  &n, 4);
+    memcpy(personalization+12, &k, 4);
+    return crypto_generichash_blake2b_init_salt_personal(&base_state,
+                                                         NULL, 0, // No key.
+                                                         n/8,
+                                                         NULL,    // No salt.
+                                                         personalization);
+}
+
+StepRow::StepRow(unsigned int n, const eh_HashState& base_state, eh_index i) :
+        hash {new unsigned char[n/8]},
+        len {n/8},
+        indices {i}
+{
+    eh_HashState state;
+    state = base_state;
+    crypto_generichash_blake2b_update(&state, (unsigned char*) &i, sizeof(eh_index));
```

Comment:
NACK:

This defines the PoW by hashing the byte layout of an ``unsigned int``, so the definition is platform specific. 

In particular this is sensitive on the size and memory endianness of ``eh_index``. Let's specify those explicitly.